src/cli.js

```diff
@@ -1,19 +1,42 @@
+const path = require('path');
 const build = require('./commands/build.js');
 const serve = require('./commands/serve.js');
 const program = require('commander');
 
 const { getBaseProgramInfo } = require('./utils/build-utils');
-const { boldGreen, boldRed, grey } = require('./utils/helpers.js');
+const { grey, boldRed, boldGreen } = require('./utils/helpers.js');
 
 program.version(require('../package.json').version);
 
-program.command('build').action(() => {
+program.command('build').action(async () => {
   try {
     const buildStartTime = new Date().getTime();
-    const programInfo = getBaseProgramInfo();
+
+    let programInfo = getBaseProgramInfo();
+
+    /** Before Build plugins */
+    for await (const pluginPath of programInfo.abellConfigs.plugins) {
+      const currentPlugin = require(pluginPath);
+      if (currentPlugin.beforeBuild) {
+        console.log(
+          '>> Plugin BeforeBuild: Executing ' +
+            path.relative(process.cwd(), pluginPath)
+        );
+        await currentPlugin.beforeBuild(programInfo);
+      }
+    }
+
+    /**
+     getBaseProgram again after making sure
+     async operations are done and "contents"  
+     folder have all the data
+    */
+    programInfo = getBaseProgramInfo();
+
```

Comment:
As you can see i have moved the beforeBuild logic in here. 

on line 15: i get the programInfo that has all the info except the contentDirectories.

Then our plugin runs and it awaits until the async operations are done.

on line 34: i again called getBaseInfoProgram to get the latest info which will include the contentDirectories and after that we will just call the build()

Reply:
Yes makes sense, `loadContent` is supposed to work individually though, what issue you faced in it? you have to pass contentPath from abellConfigs
```js
loadContent(programInfo.abellConfigs.contentPath)
```

Reply:
There was some $path is undefined error.... let me check that again

Reply:
TypeError: Cannot read property '$path' of undefined
    at generateContentFile (D:\CAM Files\Dreamwaver\Javascript\github\abell\src\utils\build-utils.js:265:31)
    at build (D:\CAM Files\Dreamwaver\Javascript\github\abell\src\commands\build.js:40:9)
    at Command.<anonymous> (D:\CAM Files\Dreamwaver\Javascript\github\abell\src\cli.js:40:5)

Reply:
And btw its happening only when there is no content & dist folder

Reply:
yes there is supposed to be a content folder for it to work,

Reply:
loadContent will basically look into all meta.json files inside `content` folder